engine/src/main/java/org/pentaho/di/job/Job.java

```diff
@@ -0,0 +1,2367 @@
+    }
+
+    JobExecutionExtension extension = new JobExecutionExtension( this, result, null, false );
+    ExtensionPointHandler.callExtensionPoint( log, KettleExtensionPoint.JobBeginProcessing.id, extension );
+
+    return true;
+  }
+
+  //
+  // Handle logging at end
+  /**
+   * End processing.
+   *
+   * @return true, if successful
+   * @throws KettleJobException
+   *           the kettle job exception
+   */
+  private boolean endProcessing() throws KettleJobException {
+    LogStatus status;
+    if ( !isActive() ) {
+      if ( isStopped() ) {
+        status = LogStatus.STOP;
+      } else {
+        status = LogStatus.END;
+      }
+    } else {
+      status = LogStatus.RUNNING;
+    }
+    try {
+      if ( errors.get() == 0 && result != null && !result.getResult() ) {
+        errors.incrementAndGet();
+      }
+
+      logDate = new Date();
+
+      /*
+       * Sums errors, read, written, etc.
+       */
+
+      JobLogTable jobLogTable = jobMeta.getJobLogTable();
+      if ( jobLogTable.isDefined() ) {
+
+        writeLogTableInformation( jobLogTable, status );
+      }
+
+      return true;
+    } catch ( Exception e ) {
+      throw new KettleJobException( e ); // In case something else goes wrong.
+    }
+  }
+
+  /**
+   * Writes information to Job Log table. Cleans old records, in case job is finished.
+   */
+  protected void writeLogTableInformation( JobLogTable jobLogTable, LogStatus status ) throws KettleJobException,
+    KettleDatabaseException {
+    boolean cleanLogRecords = status.equals( LogStatus.END );
+    String tableName = jobLogTable.getActualTableName();
+    DatabaseMeta logcon = jobLogTable.getDatabaseMeta();
+
+    Database ldb = createDataBase( logcon );
+    ldb.shareVariablesWith( this );
+    try {
+      ldb.connect();
+      ldb.setCommit( logCommitSize );
+      ldb.writeLogRecord( jobLogTable, status, this, null );
+
+      if ( cleanLogRecords ) {
+        ldb.cleanupLogRecords( jobLogTable, getJobname() );
+      }
+
+    } catch ( KettleDatabaseException dbe ) {
+      addErrors( 1 );
+      throw new KettleJobException( "Unable to end processing by writing log record to table " + tableName, dbe );
+    } finally {
+      if ( !ldb.isAutoCommit() ) {
+        ldb.commitLog( true, jobLogTable );
+      }
+      ldb.disconnect();
+    }
+  }
+
+  /**
+   * Write log channel information.
+   *
+   * @throws KettleException
+   *           the kettle exception
+   */
+  protected void writeLogChannelInformation() throws KettleException {
+    Database db = null;
+    ChannelLogTable channelLogTable = jobMeta.getChannelLogTable();
+
+    // PDI-7070: If parent job has the same channel logging info, don't duplicate log entries
+    Job j = getParentJob();
+
+    if ( j != null ) {
+      if ( channelLogTable.equals( j.getJobMeta().getChannelLogTable() ) ) {
+        return;
+      }
+    }
+    // end PDI-7070
+
+    try {
+      db = new Database( this, channelLogTable.getDatabaseMeta() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *db*. The resource is referenced at lines 1222, 1223, 1224, 1228, 1233, 1239, 1240. The resource is closed at line 1242. There are other execution paths that do not contain closure statements, for example, when Database.commit throws an exception, and an if condition at line 1239 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).